chaos/chaos-server.go

```diff
@@ -47,14 +50,42 @@ func startRPCServer(port int) {
 		lockMap: make(map[string][]lockRequesterInfo),
 		// timestamp: leave uninitialized for testing (set to real timestamp for actual usage)
 	}
-	go func() {
+	lockHeartBeat := uint64(0)
+
+	lockMaintainer := func() {
 		// Start with random sleep time, so as to avoid "synchronous checks" between servers
 		time.Sleep(time.Duration(rand.Float64() * float64(LockMaintenanceLoop)))
 		for {
 			time.Sleep(LockMaintenanceLoop)
 			locker.lockMaintenance(LockCheckValidityInterval)
+			atomic.AddUint64(&lockHeartBeat, 1)
 		}
-	}()
+	}
+	lockMonitor := func() {
+		lastHeartBeat := uint64(0)
+		lastHeartBeatStamp := time.Now()
+		for {
+			time.Sleep(LockMaintenanceLoop)
+			currentHeartBeat := atomic.LoadUint64(&lockHeartBeat)
+			if currentHeartBeat > lastHeartBeat {
+				lastHeartBeat = currentHeartBeat
+				lastHeartBeatStamp = time.Now()
+			}
+			// Check whether not too much time has passed since last 'sign' of life
+			// from maintenance loop
+			if time.Since(lastHeartBeatStamp) > LockMaintenanceHealthThreshold {
```

Comment:
The idea of the lockMonitor go-routine is to 'monitor' the health of the lockMaintenance thread, so when this thread stops running (due to any unexpected exception, memory exception, or whatever), then the `lockHeartBeat` counter will stop incrementing.
When the elapsed time between the last time that a heartbeat increased was detected and the actual time is more than 10x (or whatever you want to set it to) the LockMaintenance loop delay, then it is assumed that the lockMaintainer thread is no longer running and it is the restarted.
Note that the execution of the lockMaintenance "worker" function should be small in comparison with the LockMaintenance loop, so for production environment the values in lines 35-37 should be used.